Introduce generic MultiStreamGuard

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57052 Merge CUDAFuture into ivalue::Future
* #57051 Make CUDAFuture handle any kind of device type
* #57050 Use only generic helpers in CUDAFuture
* **#57049 Introduce generic MultiStreamGuard**
* #57048 Simplify CUDAMultiStreamGuard
* #57047 Add recordDataPtrOnStream to DeviceGuardImplInterface
* #57046 Add getStreamFromPool to DeviceGuardImplInterface

There was a comment above CUDAMultiStreamGuard which said "TODO: Implement this generically in c10". This is what I'm doing here.

The new generic MultiStreamGuard class is able to take a vector of device-agnostic c10::Streams and is able to support any device type (CUDA, but also ROCm and others) by using a VirtualGuardImpl. A class called CUDAMultiStreamGuard is still kept around, for convenience, and slightly for performance as it avoids a vtable lookup.

Differential Revision: [D28029158](https://our.internmc.facebook.com/intern/diff/D28029158/)

lw committed:
Introduce generic MultiStreamGuard

There was a comment above CUDAMultiStreamGuard which said "TODO: Implement this generically in c10". This is what I'm doing here.

The new generic MultiStreamGuard class is able to take a vector of device-agnostic c10::Streams and is able to support any device type (CUDA, but also ROCm and others) by using a VirtualGuardImpl. A class called CUDAMultiStreamGuard is still kept around, for convenience, and slightly for performance as it avoids a vtable lookup.

Differential Revision: [D28029158](https://our.internmc.facebook.com/intern/diff/D28029158/)

[ghstack-poisoned]

lw committed:
Update on "Introduce generic MultiStreamGuard"


There was a comment above CUDAMultiStreamGuard which said "TODO: Implement this generically in c10". This is what I'm doing here.

The new generic MultiStreamGuard class is able to take a vector of device-agnostic c10::Streams and is able to support any device type (CUDA, but also ROCm and others) by using a VirtualGuardImpl. A class called CUDAMultiStreamGuard is still kept around, for convenience, and slightly for performance as it avoids a vtable lookup.

Differential Revision: [D28029158](https://our.internmc.facebook.com/intern/diff/D28029158/)

[ghstack-poisoned]

mrshenli reviewed (commented):
This is great! I will also need this to use CUDAFuture in `to_here` RPC request callbacks

lw committed:
Update on "Introduce generic MultiStreamGuard"


There was a comment above CUDAMultiStreamGuard which said "TODO: Implement this generically in c10". This is what I'm doing here.

The new generic MultiStreamGuard class is able to take a vector of device-agnostic c10::Streams and is able to support any device type (CUDA, but also ROCm and others) by using a VirtualGuardImpl. A class called CUDAMultiStreamGuard is still kept around, for convenience, and slightly for performance as it avoids a vtable lookup.

Differential Revision: [D28029158](https://our.internmc.facebook.com/intern/diff/D28029158/)

[ghstack-poisoned]

lw committed:
Update on "Introduce generic MultiStreamGuard"


There was a comment above CUDAMultiStreamGuard which said "TODO: Implement this generically in c10". This is what I'm doing here.

The new generic MultiStreamGuard class is able to take a vector of device-agnostic c10::Streams and is able to support any device type (CUDA, but also ROCm and others) by using a VirtualGuardImpl. A class called CUDAMultiStreamGuard is still kept around, for convenience, and slightly for performance as it avoids a vtable lookup.

Differential Revision: [D28029158](https://our.internmc.facebook.com/intern/diff/D28029158/)

[ghstack-poisoned]

mrshenli commented on c10/core/impl/InlineStreamGuard.h:
```diff
@@ -184,4 +185,56 @@ class InlineOptionalStreamGuard {
+    DeviceType type = streams[0].device_type();
+    for (size_t idx = 1; idx < streams.size(); idx++) {
+      TORCH_CHECK_VALUE(
+        streams[1].device_type() == type,
```
you mean `streams[idx]`?

mrshenli commented on c10/core/impl/InlineStreamGuard.h:
```diff
@@ -184,4 +185,56 @@ class InlineOptionalStreamGuard {
+        streams[1].device_type() == type,
+        "Streams have a mix of device types: stream 0 is on ",
+        streams[0].device(), " while stream ", idx, " is on device ",
+        streams[1].device());
```
ditto

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -401,41 +399,6 @@ C10_REGISTER_CREATOR(
 
 } // namespace
 
-namespace {
```
nice!

mrshenli reviewed (approved):
LGTM! Except the above comments on `streams[idx]`

lw committed:
Update on "Introduce generic MultiStreamGuard"


There was a comment above CUDAMultiStreamGuard which said "TODO: Implement this generically in c10". This is what I'm doing here.

The new generic MultiStreamGuard class is able to take a vector of device-agnostic c10::Streams and is able to support any device type (CUDA, but also ROCm and others) by using a VirtualGuardImpl. A class called CUDAMultiStreamGuard is still kept around, for convenience, and slightly for performance as it avoids a vtable lookup.

Differential Revision: [D28029158](https://our.internmc.facebook.com/intern/diff/D28029158/)

[ghstack-poisoned]

lw committed:
Update on "Introduce generic MultiStreamGuard"


There was a comment above CUDAMultiStreamGuard which said "TODO: Implement this generically in c10". This is what I'm doing here.

The new generic MultiStreamGuard class is able to take a vector of device-agnostic c10::Streams and is able to support any device type (CUDA, but also ROCm and others) by using a VirtualGuardImpl. A class called CUDAMultiStreamGuard is still kept around, for convenience, and slightly for performance as it avoids a vtable lookup.

Differential Revision: [D28029158](https://our.internmc.facebook.com/intern/diff/D28029158/)

[ghstack-poisoned]

lw committed:
Update on "Introduce generic MultiStreamGuard"


There was a comment above CUDAMultiStreamGuard which said "TODO: Implement this generically in c10". This is what I'm doing here.

The new generic MultiStreamGuard class is able to take a vector of device-agnostic c10::Streams and is able to support any device type (CUDA, but also ROCm and others) by using a VirtualGuardImpl. A class called CUDAMultiStreamGuard is still kept around, for convenience, and slightly for performance as it avoids a vtable lookup.

Differential Revision: [D28029158](https://our.internmc.facebook.com/intern/diff/D28029158/)

[ghstack-poisoned]

lw committed:
Update on "Introduce generic MultiStreamGuard"


There was a comment above CUDAMultiStreamGuard which said "TODO: Implement this generically in c10". This is what I'm doing here.

The new generic MultiStreamGuard class is able to take a vector of device-agnostic c10::Streams and is able to support any device type (CUDA, but also ROCm and others) by using a VirtualGuardImpl. A class called CUDAMultiStreamGuard is still kept around, for convenience, and slightly for performance as it avoids a vtable lookup.

Differential Revision: [D28029158](https://our.internmc.facebook.com/intern/diff/D28029158/)

[ghstack-poisoned]
